spark/src/main/scala/geotrellis/spark/knn/KNearestRDD.scala

```diff
@@ -0,0 +1,27 @@
+package geotrellis.spark.knn
+
+import geotrellis.spark._
+import geotrellis.vector._
+import com.vividsolutions.jts.geom.Envelope
+import com.vividsolutions.jts.geom.Coordinate
+
+import org.apache.spark.rdd.RDD
+
+object KNearestRDD {
+  def kNearest[T](rdd: RDD[T], x: Double, y: Double, k: Int)(f: T => Extent): List[T] =
+    kNearest(rdd, new Envelope(new Coordinate(x, y)), k)(f)
+
+  def kNearest[T](rdd: RDD[T], p: (Double, Double), k: Int)(f: T => Extent): List[T] =
+    kNearest(rdd, new Envelope(new Coordinate(p._1, p._2)), k)(f)
+
+  /**
+   * Determines the k-nearest neighbors of an RDD of objects which can be coerced into Extents.
+   */
+  def kNearest[T](rdd: RDD[T], ex: Extent, k: Int)(f: T => Extent): List[T] = {
+    val candidates = rdd.glom.map { 
+        arr => SpatialIndex.fromExtents(arr)(f)
+      }.map (_.kNearest (ex, k)).reduce( _ ++ _ )
```

Comment:
It's unclear to me whether creating a SpatialIndex for each of the elements, and then using that spatial index to avoid doing distance calculations against some elements, will be less costly than just doing the distance calculation for each partition and reducing over them. What was the reason for going the spatial index route?